File: packages/ts/tests/integration/binding/persist.test.ts
// `readFileSync` stays sync: it runs inside `onFile` callbacks, whose public
// signature is synchronous.
import { readFileSync } from "node:fs";
import {
  mkdir,
  mkdtemp,
  readFile,
  rm,
  unlink,
  utimes,
  writeFile,
} from "node:fs/promises";
import { createRequire } from "node:module";
import { tmpdir } from "node:os";
import { dirname, join } from "node:path";
import { DirSQL } from "dirsql";
import initSqlJs from "sql.js";
import { afterEach, beforeEach, describe, expect, it } from "vitest";
import { exists } from "../../exists.js";

// Some tests corrupt dirsql's on-disk cache (`.dirsql/cache.db`) out-of-band
// to exercise the racy-window and dirsql_version-bump reconcile paths. sql.js
// (WASM SQLite) is used instead of `node:sqlite`, which only exists on Node
// 22.5+. sql.js is in-memory, so we read the cache bytes, mutate, and write
// them back. The cache uses WAL mode, so after closing the Rust connection,
// we must checkpoint the WAL to ensure the main db file is complete.
const resolveModule = createRequire(import.meta.url).resolve;
const sqlJsReady = initSqlJs({
  locateFile: (file) => join(dirname(resolveModule("sql.js")), file),
});

/** Open `.dirsql/cache.db` with sql.js, run `sql`, write the bytes back. */
async function corruptCache(cachePath: string, sql: string): Promise<void> {
  const SQL = await sqlJsReady;
  // With WAL mode, the cache has sidecar files (cache.db-wal, cache.db-shm).
  // sql.js reads the raw binary file and doesn't understand WAL mode.
  // Delete the sidecar files so sql.js reads a clean database file.
  try {
    await unlink(cachePath + "-wal");
  } catch {
    // File might not exist if WAL hasn't created sidecar yet
  }
  try {
    await unlink(cachePath + "-shm");
  } catch {
    // File might not exist if WAL hasn't created sidecar yet
  }

  const db = new SQL.Database(await readFile(cachePath));
  try {
    db.run(sql);
    await writeFile(cachePath, db.export());
  } finally {
    db.close();
  }
}

describe("DirSQL persist", () => {
  let dir: string;

  beforeEach(async () => {
    dir = await mkdtemp(join(tmpdir(), "dirsql-persist-"));
    await mkdir(join(dir, "items"), { recursive: true });
    await writeFile(
      join(dir, "items", "a.json"),
      JSON.stringify({ name: "apple", price: 1.5 }),
    );
  });

  afterEach(async () => {
    await rm(dir, { recursive: true, force: true });
  });

  function makeTable(box: { count: number }) {
    return {
      ddl: "CREATE TABLE items (name TEXT, price REAL)",
      glob: "items/*.json",
      onFile: (filePath: string) => {
        box.count += 1;
        return [JSON.parse(readFileSync(filePath, "utf8"))];
      },
    };
  }

  it("writes the cache db to .dirsql/cache.db on cold start", async () => {
    const box = { count: 0 };
    const db = new DirSQL({
      root: dir,
      tables: [makeTable(box)],
      persist: true,
    });
    const rows = await db.query("SELECT * FROM items");
    expect(rows).toHaveLength(1);
    expect(await exists(join(dir, ".dirsql", "cache.db"))).toBe(true);
  });

  it("trusts unchanged files on warm start", async () => {
    const box1 = { count: 0 };
    const db1 = new DirSQL({
      root: dir,
      tables: [makeTable(box1)],
      persist: true,
    });
    await db1.ready;
    expect(box1.count).toBe(1);

    const box2 = { count: 0 };
    const db2 = new DirSQL({
      root: dir,
      tables: [makeTable(box2)],
      persist: true,
    });
    await db2.ready;
    expect(box2.count).toBe(0);
    const rows = await db2.query("SELECT * FROM items");
    expect(rows).toHaveLength(1);
    expect(rows[0].name).toBe("apple");
  });

  it("re-parses changed files", async () => {
    const box1 = { count: 0 };
    const db1 = new DirSQL({
      root: dir,
      tables: [makeTable(box1)],
      persist: true,
    });
    await db1.ready;

    // Bump mtime far enough into the future to escape the racy window.
    await new Promise((r) => setTimeout(r, 50));
    await writeFile(
      join(dir, "items", "a.json"),
      JSON.stringify({ name: "cherry", price: 9.99 }),
    );
    const future = new Date(Date.now() + 5000);
    await utimes(join(dir, "items", "a.json"), future, future);

    const box2 = { count: 0 };
    const db2 = new DirSQL({
      root: dir,
      tables: [makeTable(box2)],
      persist: true,
    });
    await db2.ready;
    expect(box2.count).toBe(1);
    const rows = await db2.query("SELECT * FROM items");
    expect(rows[0].name).toBe("cherry");
  });

  it("drops rows for files removed between runs", async () => {
    await writeFile(
      join(dir, "items", "b.json"),
      JSON.stringify({ name: "banana", price: 0.75 }),
    );

    const box1 = { count: 0 };
    const db1 = new DirSQL({
      root: dir,
      tables: [makeTable(box1)],
      persist: true,
    });
    await db1.ready;

    await rm(join(dir, "items", "b.json"));

    const box2 = { count: 0 };
    const db2 = new DirSQL({
      root: dir,
      tables: [makeTable(box2)],
      persist: true,
    });
    await db2.ready;
    const rows = await db2.query("SELECT name FROM items ORDER BY name");
    expect(rows.map((r) => r.name)).toEqual(["apple"]);
  });

  it("ingests files added between runs", async () => {
    const box1 = { count: 0 };
    const db1 = new DirSQL({
      root: dir,
      tables: [makeTable(box1)],
      persist: true,
    });
    await db1.ready;

    await writeFile(
      join(dir, "items", "b.json"),
      JSON.stringify({ name: "banana", price: 0.75 }),
    );

    const box2 = { count: 0 };
    const db2 = new DirSQL({
      root: dir,
      tables: [makeTable(box2)],
      persist: true,
    });
    await db2.ready;
    expect(box2.count).toBe(1);
    const rows = await db2.query("SELECT name FROM items ORDER BY name");
    expect(rows.map((r) => r.name)).toEqual(["apple", "banana"]);
  });

  it("forces a full rebuild when the DDL changes", async () => {
    const box1 = { count: 0 };
    const db1 = new DirSQL({
      root: dir,
      tables: [makeTable(box1)],
      persist: true,
    });
    await db1.ready;

    const box2 = { count: 0 };
    const db2 = new DirSQL({
      root: dir,
      tables: [
        {
          ddl: "CREATE TABLE items (name TEXT, price REAL, sku TEXT)",
          glob: "items/*.json",
          onFile: (filePath: string) => {
            box2.count += 1;
            return [
              { ...JSON.parse(readFileSync(filePath, "utf8")), sku: "X" },
            ];
          },
        },
      ],
      persist: true,
    });
    await db2.ready;
    expect(box2.count).toBe(1);
    const rows = await db2.query("SELECT * FROM items");
    expect(rows[0].sku).toBe("X");
  });

  it("never indexes files inside the .dirsql directory", async () => {
    await mkdir(join(dir, ".dirsql", "items"), { recursive: true });
    await writeFile(
      join(dir, ".dirsql", "items", "boom.json"),
      JSON.stringify({ name: "BOOM", price: -1 }),
    );

    const db = new DirSQL({
      root: dir,
      tables: [
        {
          ddl: "CREATE TABLE items (name TEXT, price REAL)",
          glob: "**/*.json",
          onFile: (filePath: string) => [
            JSON.parse(readFileSync(filePath, "utf8")),
          ],
        },
      ],
      persist: true,
    });
    await db.ready;
    const rows = await db.query("SELECT name FROM items");
    expect(rows.map((r) => r.name)).toEqual(["apple"]);
  });

  it("hash-confirms files that fall inside the racy window", async () => {
    // Files whose cached mtime >= snapshot_ns are considered "racy" and must
    // be hash-confirmed instead of trusted. Corrupt the cached hash so the
    // hash check fails; the file must then be re-parsed.
    const box1 = { count: 0 };
    const db1 = new DirSQL({
      root: dir,
      tables: [makeTable(box1)],
      persist: true,
    });
    await db1.ready;
    expect(box1.count).toBe(1);
    // Explicitly close the connection to trigger WAL checkpoint.
    // napi close() consumes self, so this is the only call on this instance.
    db1.close();
    // Wait for WAL checkpoint to complete before reading with sql.js.
    await new Promise((resolve) => setTimeout(resolve, 50));

    const cache = join(dir, ".dirsql", "cache.db");
    await corruptCache(
      cache,
      "UPDATE _dirsql_files SET snapshot_ns = 0, content_hash = zeroblob(32)",
    );

    const box2 = { count: 0 };
    const db2 = new DirSQL({
      root: dir,
      tables: [makeTable(box2)],
      persist: true,
    });
    await db2.ready;
    expect(box2.count).toBe(1);
    const rows = await db2.query("SELECT name FROM items");
    expect(rows[0].name).toBe("apple");
  });

  it("rebuilds the cache when the dirsql_version meta changes", async () => {
    const box1 = { count: 0 };
    const db1 = new DirSQL({
      root: dir,
      tables: [makeTable(box1)],
      persist: true,
    });
    await db1.ready;
    expect(box1.count).toBe(1);
    // Explicitly close the connection to trigger WAL checkpoint.
    // napi close() consumes self, so this is the only call on this instance.
    db1.close();
    // Wait for WAL checkpoint to complete before reading with sql.js.
    await new Promise((resolve) => setTimeout(resolve, 50));

    const cache = join(dir, ".dirsql", "cache.db");
    await corruptCache(
      cache,
      "UPDATE _dirsql_meta SET value = 'bogus-version' WHERE key = 'dirsql_version'",
    );

    const box2 = { count: 0 };
    const db2 = new DirSQL({
      root: dir,
      tables: [makeTable(box2)],
      persist: true,
    });
    await db2.ready;
    expect(box2.count).toBe(1);
  });

  it("honors a custom persistPath", async () => {
    const custom = join(dir, "elsewhere", "my-cache.sqlite");
    const box = { count: 0 };
    const db = new DirSQL({
      root: dir,
      tables: [makeTable(box)],
      persist: true,
      persistPath: custom,
    });
    await db.ready;
    expect(await exists(custom)).toBe(true);
    expect(await exists(join(dir, ".dirsql", "cache.db"))).toBe(false);
  });
});

File: packages/python/tests/integration/binding/persist_test.py
"""Binding-tier tests (real core, real fs) for DirSQL persistent on-disk cache."""

import json
import os
import sqlite3
import tempfile

import pytest

from dirsql import DirSQL, Table


def _write(path, content):
    os.makedirs(os.path.dirname(path), exist_ok=True)
    with open(path, "w") as f:
        f.write(content)


def _items_table(call_count_box):
    """Build an items table whose on_file callback bumps `call_count_box[0]` per call."""

    def on_file(path):
        call_count_box[0] += 1
        return [json.loads(open(path, encoding="utf-8").read())]

    return Table(
        ddl="CREATE TABLE items (name TEXT, price REAL)",
        glob="items/*.json",
        on_file=on_file,
    )


@pytest.fixture
def persist_dir():
    with tempfile.TemporaryDirectory() as d:
        yield d


def describe_persist():
    def describe_cold_start():
        @pytest.mark.asyncio
        async def it_writes_cache_to_dotdirsql(persist_dir):
            _write(
                os.path.join(persist_dir, "items", "a.json"),
                json.dumps({"name": "apple", "price": 1.5}),
            )
            box = [0]
            db = DirSQL(persist_dir, tables=[_items_table(box)], persist=True)
            await db.ready()
            results = await db.query("SELECT * FROM items")
            assert len(results) == 1
            assert os.path.exists(os.path.join(persist_dir, ".dirsql", "cache.db"))

    def describe_warm_start():
        @pytest.mark.asyncio
        async def it_trusts_unchanged_files(persist_dir):
            _write(
                os.path.join(persist_dir, "items", "a.json"),
                json.dumps({"name": "apple", "price": 1.5}),
            )

            box1 = [0]
            db1 = DirSQL(persist_dir, tables=[_items_table(box1)], persist=True)
            await db1.ready()
            assert box1[0] == 1

            box2 = [0]
            db2 = DirSQL(persist_dir, tables=[_items_table(box2)], persist=True)
            await db2.ready()
            # Warm start: on_file not invoked for the unchanged file.
            assert box2[0] == 0
            results = await db2.query("SELECT * FROM items")
            assert len(results) == 1
            assert results[0]["name"] == "apple"

        @pytest.mark.asyncio
        async def it_leaves_the_cache_file_untouched(persist_dir):
            """An unchanged tree is a no-op: the cache is read, never rewritten."""
            _write(
                os.path.join(persist_dir, "items", "a.json"),
                json.dumps({"name": "apple", "price": 1.5}),
            )

            box1 = [0]
            db1 = DirSQL(persist_dir, tables=[_items_table(box1)], persist=True)
            await db1.ready()
            del db1

            cache = os.path.join(persist_dir, ".dirsql", "cache.db")
            before = open(cache, "rb").read()

            box2 = [0]
            db2 = DirSQL(persist_dir, tables=[_items_table(box2)], persist=True)
            await db2.ready()
            del db2

            after = open(cache, "rb").read()
            assert len(after) == len(before), "an unchanged tree must not grow the cache"
            assert after == before, "an unchanged tree must not rewrite the cache"

    def describe_changed_file():
        @pytest.mark.asyncio
        async def it_reparses_changed_files(persist_dir):
            path = os.path.join(persist_dir, "items", "a.json")
            _write(path, json.dumps({"name": "apple", "price": 1.5}))

            box1 = [0]
            db1 = DirSQL(persist_dir, tables=[_items_table(box1)], persist=True)
            await db1.ready()

            # Bump mtime far enough into the future to escape the racy window.
            import time

            time.sleep(0.05)
            _write(path, json.dumps({"name": "cherry", "price": 9.99}))
            future = time.time() + 5
            os.utime(path, (future, future))

            box2 = [0]
            db2 = DirSQL(persist_dir, tables=[_items_table(box2)], persist=True)
            await db2.ready()
            assert box2[0] == 1
            results = await db2.query("SELECT * FROM items")
            assert len(results) == 1
            assert results[0]["name"] == "cherry"

    def describe_deleted_file():
        @pytest.mark.asyncio
        async def it_drops_rows_for_deleted_files(persist_dir):
            a = os.path.join(persist_dir, "items", "a.json")
            b = os.path.join(persist_dir, "items", "b.json")
            _write(a, json.dumps({"name": "apple", "price": 1.5}))
            _write(b, json.dumps({"name": "banana", "price": 0.75}))

            box1 = [0]
            db1 = DirSQL(persist_dir, tables=[_items_table(box1)], persist=True)
            await db1.ready()

            os.remove(b)

            box2 = [0]
            db2 = DirSQL(persist_dir, tables=[_items_table(box2)], persist=True)
            await db2.ready()
            results = await db2.query("SELECT name FROM items")
            assert {r["name"] for r in results} == {"apple"}

    def describe_new_file():
        @pytest.mark.asyncio
        async def it_ingests_new_files(persist_dir):
            _write(
                os.path.join(persist_dir, "items", "a.json"),
                json.dumps({"name": "apple", "price": 1.5}),
            )

            box1 = [0]
            db1 = DirSQL(persist_dir, tables=[_items_table(box1)], persist=True)
            await db1.ready()

            _write(
                os.path.join(persist_dir, "items", "b.json"),
                json.dumps({"name": "banana", "price": 0.75}),
            )

            box2 = [0]
            db2 = DirSQL(persist_dir, tables=[_items_table(box2)], persist=True)
            await db2.ready()
            assert box2[0] == 1
            results = await db2.query("SELECT name FROM items ORDER BY name")
            assert [r["name"] for r in results] == ["apple", "banana"]

    def describe_glob_change():
        @pytest.mark.asyncio
        async def it_forces_full_rebuild_on_config_change(persist_dir):
            _write(
                os.path.join(persist_dir, "items", "a.json"),
                json.dumps({"name": "apple", "price": 1.5}),
            )

            box1 = [0]
            db1 = DirSQL(persist_dir, tables=[_items_table(box1)], persist=True)
            await db1.ready()

            # Change the DDL — this changes the glob_config_hash and forces a
            # full rebuild on the next startup.
            box2 = [0]
            db2 = DirSQL(
                persist_dir,
                tables=[
                    Table(
                        ddl="CREATE TABLE items (name TEXT, price REAL, sku TEXT)",
                        glob="items/*.json",
                        on_file=lambda path: (
                            box2.__setitem__(0, box2[0] + 1)
                            or [
                                {
                                    **json.loads(open(path, encoding="utf-8").read()),
                                    "sku": "X",
                                }
                            ]
                        ),
                    )
                ],
                persist=True,
            )
            await db2.ready()
            assert box2[0] == 1
            results = await db2.query("SELECT * FROM items")
            assert results[0]["sku"] == "X"

    def describe_dirsql_excluded():
        @pytest.mark.asyncio
        async def it_excludes_dotdirsql_from_walk(persist_dir):
            _write(
                os.path.join(persist_dir, "items", "a.json"),
                json.dumps({"name": "apple", "price": 1.5}),
            )
            # A bogus file inside .dirsql that would otherwise match the glob
            # if the scanner walked into it:
            _write(
                os.path.join(persist_dir, ".dirsql", "items", "boom.json"),
                json.dumps({"name": "BOOM", "price": -1}),
            )

            db = DirSQL(
                persist_dir,
                tables=[
                    Table(
                        ddl="CREATE TABLE items (name TEXT, price REAL)",
                        glob="**/*.json",
                        on_file=lambda path: [
                            json.loads(open(path, encoding="utf-8").read())
                        ],
                    )
                ],
                persist=True,
            )
            await db.ready()
            results = await db.query("SELECT name FROM items")
            assert {r["name"] for r in results} == {"apple"}

    def describe_racy_window():
        @pytest.mark.asyncio
        async def it_hash_confirms_files_inside_racy_window(persist_dir):
            """When a cached file's mtime falls inside the racy window
            (mtime >= snapshot_ns), the reconcile must fall back to a content
            hash instead of trusting the stat tuple. Corrupt the cached hash
            so the hash check fails; the file must be re-parsed."""
            path = os.path.join(persist_dir, "items", "a.json")
            _write(path, json.dumps({"name": "apple", "price": 1.5}))

            box1 = [0]
            db1 = DirSQL(persist_dir, tables=[_items_table(box1)], persist=True)
            await db1.ready()
            assert box1[0] == 1
            del db1  # release any file handles before mutating cache.db

            cache = os.path.join(persist_dir, ".dirsql", "cache.db")
            conn = sqlite3.connect(cache)
            # Force this file into the racy window by zeroing snapshot_ns,
            # and corrupt its cached hash so the hash-confirm branch fails.
            conn.execute(
                "UPDATE _dirsql_files SET snapshot_ns = 0, content_hash = zeroblob(32)"
            )
            conn.commit()
            conn.close()

            box2 = [0]
            db2 = DirSQL(persist_dir, tables=[_items_table(box2)], persist=True)
            await db2.ready()
            assert box2[0] == 1
            results = await db2.query("SELECT name FROM items")
            assert results[0]["name"] == "apple"

    def describe_dirsql_version_bump():
        @pytest.mark.asyncio
        async def it_rebuilds_cache_when_dirsql_version_changes(persist_dir):
            _write(
                os.path.join(persist_dir, "items", "a.json"),
                json.dumps({"name": "apple", "price": 1.5}),
            )

            box1 = [0]
            db1 = DirSQL(persist_dir, tables=[_items_table(box1)], persist=True)
            await db1.ready()
            assert box1[0] == 1
            del db1

            cache = os.path.join(persist_dir, ".dirsql", "cache.db")
            conn = sqlite3.connect(cache)
            conn.execute(
                "UPDATE _dirsql_meta SET value = 'bogus-version' "
                "WHERE key = 'dirsql_version'"
            )
            conn.commit()
            conn.close()

            box2 = [0]
            db2 = DirSQL(persist_dir, tables=[_items_table(box2)], persist=True)
            await db2.ready()
            # Version mismatch forces a full rebuild; the file is re-parsed.
            assert box2[0] == 1

    def describe_custom_persist_path():
        @pytest.mark.asyncio
        async def it_honors_custom_persist_path(persist_dir):
            _write(
                os.path.join(persist_dir, "items", "a.json"),
                json.dumps({"name": "apple", "price": 1.5}),
            )
            custom = os.path.join(persist_dir, "elsewhere", "my-cache.sqlite")
            box = [0]
            db = DirSQL(
                persist_dir,
                tables=[_items_table(box)],
                persist=True,
                persist_path=custom,
            )
            await db.ready()
            assert os.path.exists(custom)
            assert not os.path.exists(os.path.join(persist_dir, ".dirsql", "cache.db"))
